src/checker/bfs.rs

```diff
@@ -15,6 +15,13 @@ use std::sync::atomic::{AtomicUsize, Ordering};
 // While this file is currently quite similar to dfs.rs, a refactoring to lift shared
 // behavior is being postponed until DPOR is implemented.
 
+#[derive(Default)]
+struct NodeData {
+    in_degree: usize,
+    out_degree: usize,
+    parent_fingerprint: Option<Fingerprint>,
+}
```

Comment:
One thing impacting performance is the change from storing `Option<Fingerprint>` in the map (which encodes to a single `u64` as `Fingerprint` is a `NonZeroU64` enabling a memory layout optimization) to this new structure that cannot encode as efficiently.

I don't know whether it would help or hurt performance, but perhaps adopting different maps along the lines of https://en.wikipedia.org/wiki/Data-oriented_design is worth exploring.